### PR TITLE
feat: Added 1d+2d Perlin and Simplex noise functions

### PR DESCRIPTION
**Pull request by st-pasha**
# Description

These functions are used in games for a variety of purposes: terrain generation, visual effects, randomized motion, textures, enemy spawns, etc, etc.

This implementation is ported from the original code by Stefan Gustavson, and manually verified that it produces the same values as his C code. The original library had more varieties of noise functions, including 3D- and 4D- noise, as well as noise with gradients and periodic noise. If needed, those other functions can be ported as well.

## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have read the [Contributor Guide] and followed the process outlined for submitting PRs.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- (NA) I have updated/added relevant examples in `examples`.

## Breaking Change

- [x] No, this is *not* a breaking change.

## Related Issues

WIP for #396

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md
[Conventional Commit]: https://conventionalcommits.org

### CODE REVIEW

File: packages/flame/lib/algos.dart [hunk 1]
```diff
@@ -0,0 +1 @@
+export 'src/noise.dart' show noise1, noise2, snoise1, snoise2;
```
Comment (spydon) [inline packages/flame/lib/algos.dart:1]: The name of this file should be the full word.

File: packages/flame/lib/src/noise.dart [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+  final nx10 = _grad2(_perm[ix1 + _perm[iy0]], fx1, fy0);
+  final nx11 = _grad2(_perm[ix1 + _perm[iy1]], fx1, fy1);
+  final n1 = _lerp(t, nx10, nx11);
+  return 0.507 * _lerp(s, n0, n1);
+}
+
+//#endregion
+
+//#region Simplex noise
+// This is a clean, fast, modern and free Perlin Simplex noise function.
```
Comment (spydon) [inline packages/flame/lib/src/noise.dart:59]: Isn't this dartdoc?

**Review by spydon — COMMENTED** [review]
Maybe create a new directory called algorithms or something? Or we could do that when there are more of them instead.

File: packages/flame/lib/src/noise.dart (hunk 2 above)
Comment (st-pasha) [reply packages/flame/lib/src/noise.dart]: I don't know, this is a comment that relates to all functions in that section, and it was copied from the original Gustavson's implementation

Comment (spydon) [reply packages/flame/lib/src/noise.dart]: Pretty funny that it says modern since the comment was written like 20 years ago 😄

**Review by spydon — APPROVED** [review]

**Comment by spydon** [comment]
On a second thought; do we really gain anything on maintaining this inside the engine, since there are already good dart libraries for this? Example: https://pub.dev/packages/fast_noise

**Comment by st-pasha** [comment]
They don't have 1d noise

**Comment by wolfenrain** [comment]
As there are already other packages who provide this I dont see why we should also maintain another version. I have used the one @spydon mentioned in Flame many times both for 2d and 1d noise (you can just hardcode one of the files). 

And even if there were no, or bad, packages I dont think this should be a core flame feature. We should then opt for just having a perlin-like package. It has zero connections to Flame itself and could be useful outside of the flame ecosystem as well.

**Comment by st-pasha** [comment]
> you can just hardcode one of the files

This is exactly what this pr is doing

**Comment by st-pasha** [comment]
In order to avoid the [status quo bias](https://en.wikipedia.org/wiki/Status_quo_bias), we should ask ourselves the following: if this feature was already merged long time ago, would we be arguing in favor of removing it and replacing with an external dependency? How would we go about implementing the 1d noise functionality that we actually need?

**Comment by spydon** [comment]
> In order to avoid the [status quo bias](https://en.wikipedia.org/wiki/Status_quo_bias), we should ask ourselves the following: if this feature was already merged long time ago, would we be arguing in favor of removing it and replacing with an external dependency? How would we go about implementing the 1d noise functionality that we actually need?

The status quo bias doesn't apply here, new information was provided from DevKage that there already are great libraries for this, which I didn't know, and that's why I approved the PR at first. We don't need the 1d noise functionality for anything specific directly in the engine right? It is not hard to take just one of the dimensions from a 2d noise.

**Comment by st-pasha** [comment]
> We don't need the 1d noise functionality for anything specific directly in the engine right?

I need noise specifically in order to implement #396, and computing 2d noise then taking 1d slice of it is much less performant: just compare the amount of calculations in `noise1()` and `noise2()` functions.

**Comment by spydon** [comment]
> > We don't need the 1d noise functionality for anything specific directly in the engine right?
> 
> I need noise specifically in order to implement #396, and computing 2d noise then taking 1d slice of it is much less performant: just compare the amount of calculations in `noise1()` and `noise2()` functions.

Why can't you just use `Random()` for that?

**Comment by st-pasha** [comment]
> Why can't you just use Random() for that? 

All articles that I've seen on shaking in games recommend using some kind of noise function (perlin, simplex, value, sum of sines, etc), and warn against using pure Random() because it produces unnatural-looking effect.

**Comment by spydon** [comment]
> > Why can't you just use Random() for that? 
> 
> All articles that I've seen on shaking in games recommend using some kind of noise function (perlin, simplex, value, sum of sines, etc), and warn against using pure Random() because it produces unnatural-looking effect.

It must depend on how you use random? You could just generate a delta each tick and not the absolute position from random? 

> it produces unnatural-looking effect

I think this makes more sense for other movement effects, the shake is supposed to be disruptive anyways?

**Comment by st-pasha** [comment]
There is no single "right" way to shake things up. The shake from an earthquake is different from the shake from an explosion, which is different from the shake from a Big Guy walking, which is different from the shake from bullet recoil, etc. I imagine developers would want to experiment and see what works best in the context of their games. Our job then is to provide them with the necessary tools.

 > You could just generate a delta each tick and not the absolute position from random?

This would produce the [random walk](https://en.wikipedia.org/wiki/Random_walk) behavior. I know its mathematical properties, but I don't know how it will actually _feel_. Still, it could be just another tool in the toolbox.

**Comment by spydon** [comment]
> I imagine developers would want to experiment and see what works best in the context of their games. Our job then is to provide them with the necessary tools.

If this is for users of Flame, then this shouldn't be in the engine I think, possibly a PR (with 1d) to one of the libraries already providing this. If we need it internally in the engine we should have it though, since we don't want a new external dependency. As of now, I don't see much use for it internally if it is only for the shake effect, but maybe you have more usecases?

**Comment by st-pasha** [comment]
> If this is for users of Flame, then this shouldn't be in the engine I think

Everything in the engine is for users of Flame, isn't it? I mean, sure, perlin-noise-based shake effects aren't the only kind of shakes possible, but they are quite common ones. And giving the user ability to choose among several shake effects that better suite their needs is good.

I mean, there is no actual cost of including it into Flame -- the tree-shaking algorithm ensures that whatever functions are not used by the user will not be included into the final executable.

> but maybe you have more use cases?

Besides **shake effects**, noise can also be used in **particle systems**. Noise is also great for terrain generation and texture generation, but I'm not sure we're going to include that functionality into the engine anytime soon.

**Comment by spydon** [comment]
> Everything in the engine is for users of Flame, isn't it? 

It is in some sense, but if it is already done well by another package in the dart ecosystem which can be used by our users we shouldn't reinvent the wheel. This is why the question whether this is something that is needed internally or not is important.

> I mean, there is no actual cost of including it into Flame -- the tree-shaking algorithm ensures that whatever functions are not used by the user will not be included into the final executable.

The cost is an unnecessary maintenance cost, not a performance cost.

**Comment by st-pasha** [comment]
> The cost is an unnecessary maintenance cost

Well, this is a well-established stable algorithm. You probably won't find any other functionality that requires less maintenance than this.

**Comment by spydon** [comment]
> Well, this is a well-established stable algorithm. You probably won't find any other functionality that requires less maintenance than this.

In the future there might be more noise algorithms that people want to add, you even spoke about some yourself, and then it is better to focus our efforts on helping an already established library in the ecosystem than to duplicate the code.

**Comment by st-pasha** [comment]
> and then it is better to focus our efforts on helping an already established library in the ecosystem than to duplicate the code.

Well, there are several counterindications to this approach:
1. We don't want to burden Flame with external dependencies; so even if we helped another package, that would not be helpful to us.
2. We would have to contend with the pace of development of the other package. From what I see, `fast_noise` in its 4 years of existence have had only 1 PR and 2 issues (both still open).

**Comment by spydon** [comment]
> Well, there are several counterindications to this approach:
> 
>     1. We don't want to burden Flame with external dependencies; so even if we helped another package, that would not be helpful to us

But none of your examples so far has been anything that we plan to do within the engine, and for shaking I don't think we will need noise. So the users could just use any other library when creating the examples you had together with flame in their own games, so we will never have to add the dependency to flame.

>     2. We would have to contend with the pace of development of the other package. From what I see, `fast_noise` in its 4 years of existence have had only 1 PR and 2 issues (both still open).

We could contact the person and see what the state of things are, they have moved over to null safety this year at least. If it's not possible, we could fork it and provide it under bluefire, since they have several nice algorithms in there.

**Comment by erickzanardo** [comment]
Given that there is already a package for that, I don't see the point on adding this code source to Flame. If the lib is lacking something PRs should be added to it.

**Comment by spydon** [comment]
I'll open an issue for how we should solve this instead, and if we come to the conclusion that this is something we'd want directly in the engine we can re-open this PR.

**Pull request closed without merging**